docs/config.md
# Config variables

## Initial setup

If you want to override the default values for some config options, then the
file available in this repo (etc/ovn_k8s.conf) must be copied to the following
locations:

- on Linux:
```
/etc/openvswitch/ovn_k8s.conf
```
The following command copies the config file if it is run from inside the repo:
```
cp etc/ovn_k8s.conf /etc/openvswitch/ovn_k8s.conf
```

- on Windows:
```
C:\etc\ovn_k8s.conf
```
The following PowerShell command copies the config file if is run from inside
the repo:
```
Copy-Item ".\etc\ovn_k8s.conf" -Destination (New-Item "C:\etc" -Type container -Force)
```

## Config values

The config file contains common configuration options shared between the various
ovn-kubernetes programs (ovnkube, ovn-k8s-cni-overlay, etc).  All configuration
file options can also be specified as command-line arguments which override
config file options; see the -help output of each program for more details.

### [default] section

The following config option represents the MTU value which should be used
for the overlay networks.
```
mtu=1400
```

The following option affects only the gateway nodes. This value is used to
track connections that are initiated from the pods so that the reverse
connections go back to the pods. This represents the conntrack zone used
for the conntrack flow rules.
```
conntrack-zone=64000
```

The following option only affects ovn-controller. This is the maximum number
of milliseconds of idle time on connection to the server before sending an
inactivity probe message.  As a client connects to the server over TCP, it
may take a while for the kernel to figure out if the connection is
broken.  But the client can overcome this by periodically sending probes over
the TCP connection to make sure that the connection is up.  On the flip side,
when there are hundreds of nodes, the server can get bogged down by client
probe messages.

The default value is set as 100000ms. But can be changed with this config
option

inactivity-probe=600000

### [logging] section

The following config values control what verbosity level logging is written at
and to what file (if any).
```
loglevel=5
logfile=/var/log/ovnkube.log
```

### [cni] section

The following config values are used for the CNI plugin.
```
conf-dir=/etc/cni/net.d
plugin=ovn-k8s-cni-overlay
```

### [kubernetes] section

Kubernetes API options are stored in the following section.
```
apiserver=https://1.2.3.4:6443
token=TG9yZW0gaXBzdW0gZG9sb3Igc2l0IGFtZXQsIGNvbnNlY3RldHVyIGFkaXBpc2NpbmcgZWxpdC4gQ3JhcyBhdCB1bHRyaWNpZXMgZWxpdC4gVXQgc2l0IGFtZXQgdm9sdXRwYXQgbnVuYy4K
cacert=/etc/kubernetes/ca.crt
```

### [ovnnorth] section

This section contains the address and (if the 'ssl' method is used) certificates
needed to use the OVN northbound database API. Only the the ovn-kubernetes
master needs to specify the 'server' options.
```
address=ssl:1.2.3.4:6641
client-privkey=/path/to/private.key
client-cert=/path/to/client.crt
client-cacert=/path/to/client-ca.crt
server-privkey=/path/to/private.key
server-cert=/path/to/server.crt
server-cacert=/path/to/server-ca.crt
```

### [ovnsouth] section

This section contains the address and (if the 'ssl' method is used) certificates
needed to use the OVN southbound database API. Only the the ovn-kubernetes
master needs to specify the 'server' options.
```
address=ssl:1.2.3.4:6642
client-privkey=/path/to/private.key
client-cert=/path/to/client.crt
client-cacert=/path/to/client-ca.crt
server-privkey=/path/to/private.key
server-cert=/path/to/server.crt
server-cacert=/path/to/server-ca.crt
```


ovn_k8s/common/config.py
# Copyright (C) 2016 Nicira, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at:
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import ast
import os
import sys

from ovn_k8s.common.util import ovs_vsctl
from ovn_k8s.common.util import ovn_nbctl
from ovn_k8s.common import variables

if sys.version_info < (3, 0):
    # python 2
    import ConfigParser
else:
    # python 3
    import configparser as ConfigParser


raw_config_parser = ConfigParser.RawConfigParser()


def init_config():
    config_filename = 'ovn_k8s.conf'
    if sys.platform != 'win32':
        config_path = "/etc/openvswitch"
    else:
        config_path = "C:\\etc"
    config_file = '%s/%s' % (config_path, config_filename)
    # Convert the path to use platform specific path delimitators
    config_file = os.path.abspath(config_file)
    # Return True if the file could be read, otherwise False
    return len(raw_config_parser.read(config_file)) == 1


config_successfully_read = init_config()


def get_default_value(option_name, section_name):
    section_default = {
        "mtu": 1400,
        "conntrack_zone": 64000,
        "ovn_mode": "overlay",
        "log_path": "/var/log/openvswitch/ovn-k8s-cni-overlay.log",
        "unix_socket": "/var/run/openvswitch/ovnnb_db.sock",
        "cni_conf_path": "/etc/cni/net.d",
        "cni_link_path": "/opt/cni/bin/",
        "cni_plugin": "ovn-k8s-cni-overlay",
        "private_key": "/etc/openvswitch/ovncontroller-privkey.pem",
        "certificate": "/etc/openvswitch/ovncontroller-cert.pem",
        "ca_cert": "/etc/openvswitch/ovnnb-ca.cert",
        "k8s_ca_certificate": "/etc/openvswitch/k8s-ca.crt",
        "rundir": "",
        "logdir": ""
    }
    default_config = {'default': section_default}

    section_dict = default_config.get(section_name, {})
    return section_dict.get(option_name, None)


def get_option(option_name, section_name='default'):
    if not config_successfully_read:
        return get_default_value(option_name, section_name)

    try:
        config_string = raw_config_parser.get(section_name, option_name)
    except Exception:
        # Config value could not be found in the file, retrieve the
        # default value
        config_string = get_default_value(option_name, section_name)

    try:
        # Try to evaluate the string which may contain a Python expression
        expr = ast.literal_eval(config_string)
        return expr
    except Exception:
        return config_string


UNIX_SOCKET = get_option('unix_socket')


def ovn_init_overlay():
    if os.path.exists(UNIX_SOCKET):
        OVN_NB = "unix:%s" % UNIX_SOCKET
    else:
        sys.exit("OVN NB database does not have unix socket")
    variables.OVN_NB = OVN_NB

    K8S_API_SERVER = ovs_vsctl("--if-exists", "get", "Open_vSwitch", ".",
                               "external_ids:k8s-api-server").strip('"')
    if not K8S_API_SERVER:
        sys.exit("K8S_API_SERVER not set")
    if not K8S_API_SERVER.startswith("http"):
        variables.K8S_API_SERVER = "http://%s" % K8S_API_SERVER
    else:
        variables.K8S_API_SERVER = K8S_API_SERVER

    K8S_CLUSTER_ROUTER = ovn_nbctl("--data=bare", "--no-heading",
                                   "--columns=_uuid", "find", "logical_router",
                                   "external_ids:k8s-cluster-router=yes")
    if not K8S_CLUSTER_ROUTER:
        sys.exit("K8S_CLUSTER_ROUTER not set")
    variables.K8S_CLUSTER_ROUTER = K8S_CLUSTER_ROUTER

    K8S_CLUSTER_LB_TCP = ovn_nbctl("--data=bare", "--no-heading",
                                   "--columns=_uuid", "find", "load_balancer",
                                   "external_ids:k8s-cluster-lb-tcp=yes")
    if not K8S_CLUSTER_LB_TCP:
        sys.exit("K8S_CLUSTER_LB_TCP not set")
    variables.K8S_CLUSTER_LB_TCP = K8S_CLUSTER_LB_TCP

    K8S_CLUSTER_LB_UDP = ovn_nbctl("--data=bare", "--no-heading",
                                   "--columns=_uuid", "find", "load_balancer",
                                   "external_ids:k8s-cluster-lb-udp=yes")
    if not K8S_CLUSTER_LB_UDP:
        sys.exit("K8S_CLUSTER_LB_UDP not set")
    variables.K8S_CLUSTER_LB_UDP = K8S_CLUSTER_LB_UDP
